### PR TITLE
fix #14009; check for types growing due to recursion and widen them

### LINKED ISSUE #14009: Avoid infinite type recursion in type inference
See point 2 in https://github.com/JuliaLang/julia/issues/14005#issuecomment-156849772

A [minimum repro](https://github.com/yuyichao/explore/blob/9f1f81b453ee3c9b70a09e8e12c3bd81a990cb82/julia/wb-freeze/typeinf-recursion.jl) is,

```jl
type A{T}
end

A{T}(a::T) = A{T}()

function f(a)
    if rand() > 100
        return f(A(a))
    end
end

@code_warntype f(1)
```

Even though the recursion path will never be executed, type inference will still try to infer the type of the recursion call with a infinitely deep type and even if there's a genuine infinite recursion, it's better to do that at runtime and not in type inference.....

We already have a bunch of limit on type length so I think we should probably also have some limit here.

@carnaval Does green-fairy provide any general solution to these problems?

### PR DESCRIPTION
**Pull request by JeffBezanson**
This is an alternative to #14059 that should be a bit more optimistic.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
This should probably be checked for both run time and compiler performance regressions.

File: test/core.jl [hunk 1]
```diff
@@ -3586,3 +3586,13 @@ function __f_isa_arg_1()
     length(a)
 end
 @test __f_isa_arg_1() == 1
+
+# non-terminating inference, issue #14009
+type A14009{T}; end
+A14009{T}(a::T) = A14009{T}()
+f14009(a) = ? ? f14009(A14009(a)) : a
```
Comment (vtjnash) [inline test/core.jl]: is it really necessary to use `?` as a variable here? i used to think this was was an unintentional parser "bug"...

Comment (JeffBezanson) [inline test/core.jl]: I'll change it. I was just copying oscar's test case.

**Comment by JeffBezanson** [comment]
Ok I've looked at sysimg build time, test-pef, and test-linalg and I don't see any major differences. Sysimg build takes 10s longer, but that's ok.

**Pull request merged**